### ISSUE TITLE
Support for OpenGL Core Contexts

### ISSUE THREAD

**Issue by scott-howard-james** [vispy/vispy#952]
On OS X configurations supporting OpenGL 3.0+ get following warning when running vispy examples:

WARNING: OpenGL version 3.0 or higher recommended, got 2.1 NVIDIA-10.2.7 310.41.25f01. Some functionality may fail.

Possibly this is due to OpenGL compatibility profile being restricted to 2.1 for OS/X.  Is there a plan to support Core Profile Only modes?

Thanks

https://www.opengl.org/wiki/Core_And_Compatibility_in_Contexts
http://stackoverflow.com/questions/25708478/pyqt4-opengl-enabling-opengl-core-profile

**Comment by QuLogic** [comment]
Also, switching to Core might be an alternate way to fix #949/#954.

**Comment by Eric89GXL** [comment]
The stackoverflow makes it sound like this might be a Qt thing. Do you get this when you use other backends as well?

**Comment by scott-howard-james** [comment]
I indeed did try this under Qt4, Qt5 but also pyglet.
I also examined pyglet directly (i.e. outside vispy) and it appears to default to compatibility mode unless forced to core.

That said, this issue doesn't appear to be just a Qt thing:

http://stackoverflow.com/questions/21526662/opengl-3-3-on-osx-with-freeglut
https://github.com/RPCS3/rpcs3/issues/1028
http://you-script.org/thomasena/2015/05/05/osx-opengl-4-1-under-mavericks/

**Comment by Eric89GXL** [comment]
I think it would make sense to support enabling core profile mode. I'm not sure if we should try to do it by default though -- I'd have to think about what that would mean for supporting OpenGL ES 2.0 code. @scott-howard-james do you want to try adding an option to use core profile mode? It would probably go in the `config` dictionary that `vispy.app.Canvas` has, and then would need to be implemented for each backend (maybe just PyQt4 for now?).

**Comment by scott-howard-james** [comment]
Glad to help.  However, it might be a bit more involved than a switch. 

Following the advice in <http://stackoverflow.com/questions/25708478/pyqt4-opengl-enabling-opengl-core-profile> I modified the app/backends/_qt.py (see below) to force a core context (I also tried 4.1... the highest available on this particular laptop)

Unfortunately, this creates a runtime problem in shader code (see below).  My knowledge of OpenGL is not deep, so I'm guessing here, but this looks like an issue with mixing legacy-pre-3 code and 'modern'-post-3 code. 

A related and illuminating discussion can be found here: 

<https://groups.google.com/forum/#!topic/pyglet-users/9GMilKuFg7g>

--- snip ---

def _set_config(c):
    """Set the OpenGL configuration"""
    glformat = QGLFormat()
    glformat.setVersion(3, 3)
    glformat.setProfile(QGLFormat.CoreProfile)
...

--- snip ---

RuntimeError: Shader compilation error in vertex shader:
    on line 1: '' :  version '120' is not supported
      #version 120
    on line 2: '' :  #version required and missing.
      
    on line 12: 'attribute' : syntax error: syntax error
      attribute vec3 a_position;
ERROR: Invoking <bound method Canvas.on_draw of <Canvas (PyQt5) at 0x10f83e128>> for DrawEvent
ERROR: Invoking <bound method Canvas.on_draw of <Canvas (PyQt5) at 0x10f83e128>> repeat 2
...

**Comment by campagnola** [comment]
Right; our shaders are written for GLSL 120, which corresponds to GL 2.1 (so this should be used with the compatibility profile, not core). I do think it would be valuable, though, to begin working on (or at least enumerating) the pieces required to support core.

**Comment by Eric89GXL** [comment]
Ahh. Perhaps we should update our warning to just warn if gl < 2.1 (instead of 3.0) then. @campagnola WDYT?

**Comment by campagnola** [comment]
Agreed.

**Comment by rossant** [comment]
Adding

```
+    glformat.setVersion(4, 1)
+    glformat.setProfile(QGLFormat.CoreProfile)
```

fixes the problem for me on 0.4.0.

**Comment by Eric89GXL** [comment]
Is there an easy way to use Qt to query the max available version, and use that? Or can we just request the highest available version and it will fall back?
